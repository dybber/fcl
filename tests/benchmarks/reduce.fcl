-- Reduce 2: Assumes commutativity + associativity.
sig reduceBlock : forall <lvl> 'a. ('a -> 'a -> 'a) -> ['a] -> Program<lvl> ['a]<lvl>
fun reduceBlock<lvl> f arr =
  let cond = fn arr => 1 != lengthPull arr in
  let step = fn arr => let x = halve arr
                       in pushM<lvl> (zipWith f (fst x) (snd x))
  in do <lvl>
      { a <- while<lvl> cond step (step arr)
      ; pushM<lvl> a
      }

-- Reduce array partially by one kernel invocation
sig reducePart : forall 'a. ('a -> 'a -> 'a) -> 'a -> ['a] -> Program<grid> ['a]<grid>
fun reducePart f neutralElement arr =
 let workPerBlock = 2 * 256 in
 let paddedArray = pad workPerBlock neutralElement arr     
 in splitUp workPerBlock paddedArray
     |> concatMap<block> 1 (reduceBlock<block> f) 

-- Full reduction
sig reduce : forall 'a. ('a -> 'a -> 'a) -> 'a -> ['a] -> Program <grid> 'a
fun reduce f neutralElement arr =
  do<grid>
   { a <- while<grid>
             (fn arr => 1 != lengthPull arr)
             (reducePart f neutralElement)
             (reducePart f neutralElement arr)
   ; return<grid> (index a 0)
   }

------------------
-- Benchmarking --
------------------
sig benchPartialReduction : [int] -> Program <grid> [int]<grid>
fun benchPartialReduction input = reducePart (fn a => fn b => a + b) 0 input

sig benchFullReduction : [int] -> Program<grid> int
fun benchFullReduction input = reduce (fn a => fn b => a + b) 0 input

sig main : Program<grid> unit
val main =
  do<grid>
   { input <- generateInput (4096*4096)
   ; arr1 <- benchPartialReduction input
   ; forceAndPrint 100 arr1
   ; i <- benchFullReduction input
   ; benchmark 100 (forceM<grid> (benchPartialReduction input))
   ; benchmark 100 (benchFullReduction input)
   }
