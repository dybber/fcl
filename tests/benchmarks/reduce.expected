[130816,392960,655104,917248,1179392,1441536,1703680,1965824,2227968,2490112,2752256,3014400,3276544,3538688,3800832,4062976,4325120,4587264,4849408,5111552,5373696,5635840,5897984,6160128,6422272,6684416,6946560,7208704,7470848,7732992,7995136,8257280,8519424,8781568,9043712,9305856,9568000,9830144,10092288,10354432,10616576,10878720,11140864,11403008,11665152,11927296,12189440,12451584,12713728,12975872,13238016,13500160,13762304,14024448,14286592,14548736,14810880,15073024,15335168,15597312,15859456,16121600,16383744,16645888,16908032,17170176,17432320,17694464,17956608,18218752,18480896,18743040,19005184,19267328,19529472,19791616,20053760,20315904,20578048,20840192,21102336,21364480,21626624,21888768,22150912,22413056,22675200,22937344,23199488,23461632,23723776,23985920,24248064,24510208,24772352,25034496,25296640,25558784,25820928,26083072]
